src/linear_algebra/general_linear_group.lean

```diff
@@ -0,0 +1,267 @@
+
+
+## References
+
+
+## Tags
+
+matrix group, group, matrix inverse
+-/
+
+namespace matrix
+universes u v
+open_locale matrix
+open linear_map
+
+
+section
+
+variables (n : Type u) [decidable_eq n] [fintype n] (R : Type v) [comm_ring R]
+
+/-- `GL n R` is the group of `n` by `n` `R`-matrices with unit determinant.
+Defined as a subtype of matrices
+-/
+abbreviation general_linear_group (R : Type*) [comm_ring R] :  Type* := units (matrix n n R )
+
+notation `GL`:= general_linear_group
+
+
+
+lemma GL.is_unit_det  (A : GL n R) : is_unit ( (A : matrix n n R).det) :=
+begin
+  rw ←matrix.is_unit_iff_is_unit_det,
+  exact units.is_unit _
+end
+
+
+/--The `GL n R` and `general_linear_group R n` are multiplicatively equivalent-/
+def GL_mul_equiv_general_linear_group : (GL n R) ≃* (linear_map.general_linear_group R ( n → R)) :=
+units.map_equiv to_lin_alg_equiv'.to_mul_equiv
+
+
+/--Given a matrix with invertible determinant we get an element of `GL n R`-/
+ def GL.mk' (A: matrix n n R) (h: invertible (det A)): GL n R:=
+unit_of_det_invertible A
+
+/--Given a matrix with unit determinant we get an element of `GL n R`-/
+noncomputable def GL.mk'' (A: matrix n n R) (h: is_unit (det A)): GL n R:=
+nonsing_inv_unit A h
+
+
+end
+namespace general_linear_group
+
+
+
+variables {n : Type u} [decidable_eq n] [fintype n] {R : Type v} [comm_ring R]
+
+instance coe_fun : has_coe_to_fun (GL n R) :=
+{ F   := λ _, n → n → R,
+  coe := λ A, A.val }
+
+/--Given a element of `GL n R` it gives the associated linear map-/
+
+def to_lin' (A : GL n R) := matrix.to_lin' A
+
+lemma ext_iff (A B : GL n R) : A = B ↔ (∀ i j, A i j = B i j) :=
+begin
+dsimp at *, fsplit, work_on_goal 0 { intros a i j, induction a, refl }, intros a,
+ ext1, dsimp at *, ext1, solve_by_elim,
+end
+
+@[ext] lemma ext (A B : GL n R) : (∀ i j, A i j = B i j) → A = B :=
+(ext_iff A B).mpr
+
+
+
+section coe_lemmas
+
+variables (A B : GL n R)
+
+
+
+
+
+@[simp] lemma mul_val : ↑(A * B) = A ⬝ B := rfl
+
+@[simp] lemma mul_apply : ⇑(A * B) = (A ⬝ B) := rfl
+
+
+
+
+@[simp] lemma one_val : ↑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+@[simp] lemma one_apply : ⇑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+
+@[simp] lemma to_lin'_mul :
+  to_lin' (A * B) = (to_lin' A).comp (to_lin' B) :=
+matrix.to_lin'_mul A B
+
+@[simp] lemma to_lin'_one :
+  to_lin' (1 : GL n R) = linear_map.id :=
+matrix.to_lin'_one
+
+
+
+
+noncomputable instance :  has_inv (GL n R):=
+⟨λ A, ⟨nonsing_inv A, A, by {simp, apply nonsing_inv_mul, apply GL.is_unit_det _ _ A,},
+by { simp, apply mul_nonsing_inv, apply GL.is_unit_det _ _ A, }, ⟩⟩
+
+@[simp] lemma inv_val : ↑(A⁻¹) = nonsing_inv A := rfl
+
+@[simp] lemma inv_apply : ⇑(A⁻¹) = nonsing_inv A := rfl
+
+end coe_lemmas
+
+lemma is_left_inv (A : GL n R): A⁻¹ *  A = 1:=
+
+begin
+ have h0:= GL.is_unit_det _ _ A,
+have:=nonsing_inv_mul A h0, rw ← mul_eq_mul at this, ext, dsimp,rw nonsing_inv,
+cases h0, dsimp at *, solve_by_elim,
+end
+
+
+noncomputable instance : group (GL n R) :=
```

Comment:
Doesn't this instance exists already?

Reply:
So yes, I just used the definition of the group structure for units in general, but I wanted to make the inverse more explicit, since it kept braking my `inv_det_pos` and `GL_pos` definitions. Using the inverse coming from infering this group instance was causing me trouble so I thought I'd do this.

Reply:
You'd need to remove your custom `has_inv` too

Reply:
Yes so doing it without the custom group definition `inv_det_pos` ends up asking me to prove  `A^-1.val=(A.val)^-1` (sorry I don't know how to do lean code here)  where A is an element of `GL n R`. Now, I tried proving this, but it was tricky and couldn't do it. But that is likely because I'm not very good at lean.

Reply:
If you end up with a goal you can't prove, you can always leave a `sorry` and ask reviewers (or zulip) for help :)

Reply:
Would you mind switching back to the group instance that already exists, and leaving a sorry where you get stuck?

Reply:
Aha, I was thinking that one wouldn't want sorrys in the commits, but that is good to know. I've switched it now and pushed the sorry version.

Reply:
As long as the sorry doesn't get merged it's fine. Bors will squash your commits and there will be no record of the sorry in master if this gets merged.